src/main/java/com/vmware/vipclient/i18n/messages/api/opt/server/RemoteLocaleOpt.java

```diff
@@ -30,61 +30,129 @@ public RemoteLocaleOpt(LocaleDTO dto) {
         logger.debug("Look for regions from Singleton Service for locale [{}]!", locale);
-    	Map<String, Object> response = VIPCfg.getInstance().getVipService().getHttpRequester().request(
-                V2URL.getRegionListURL(locale, VIPCfg.getInstance().getVipService().getHttpRequester().getBaseURL()),
-                ConstantsKeys.GET, null);
-    	String responseData = (String) response.get(URLUtils.BODY);
-        if(responseData == null || responseData.isEmpty())
-            return null;
-        Map<String, String> respMap = null;
         try {
-            JSONObject jsonObject = (JSONObject) JSONValue.parseWithException(responseData);
-            JSONArray jsonArray = (JSONArray) jsonObject.get(ConstantsKeys.DATA);
-            if (jsonArray != null && !jsonArray.isEmpty()) {
-                Map<String, Object> regionMap = JSONUtils.getMapFromJson(jsonArray.get(0).toString());
-                respMap = (Map<String, String>) regionMap.get(ConstantsKeys.TERRITORIES);
+            Map<String, Object> response = getResponse(
+                    V2URL.getRegionListURL(locale, VIPCfg.getInstance().getVipService().getHttpRequester().getBaseURL()),
+                    ConstantsKeys.GET, null, cacheItem);
+
+            Integer responseCode = (Integer) response.get(URLUtils.RESPONSE_CODE);
+
+            if (responseCode != null && (responseCode.equals(HttpURLConnection.HTTP_OK) ||
+                    responseCode.equals(HttpURLConnection.HTTP_NOT_MODIFIED))) {
+                long timestamp = 0;
+                String etag = null;
+                Long maxAgeMillis = null;
+                if (response.get(URLUtils.RESPONSE_TIMESTAMP) != null)
+                    timestamp = (long) response.get(URLUtils.RESPONSE_TIMESTAMP);
+                if (response.get(URLUtils.HEADERS) != null)
+                    etag = URLUtils.createEtagString((Map<String, List<String>>) response.get(URLUtils.HEADERS));
+                if (response.get(URLUtils.MAX_AGE_MILLIS) != null)
+                    maxAgeMillis = (Long) response.get(URLUtils.MAX_AGE_MILLIS);
+
+                if (responseCode.equals(HttpURLConnection.HTTP_OK)) {
+                    try {
+                        String responseBody = (String) response.get(URLUtils.BODY);
+                        Map<String, String> territories = getTerritoriesFromResponse(responseBody);
+                        if (territories != null) {
+                            logger.debug("Find the regions from Singleton Service for locale [{}].\n", locale);
+                            territories = JSONUtils.map2SortMap(territories);
+                            cacheItem.set(territories, etag, timestamp, maxAgeMillis);
+                        } else {
+                            logger.debug("Doesn't find the regions from Singleton Service for locale [{}].\n", locale);
+                            cacheItem.set(etag, timestamp, maxAgeMillis);
```

Comment:
This is for handling non-supported locales, just want to add an empty fetch result to cache, it's the same as you new  an empty MessageCacheItem and add it to cache.

Reply:
_"The cacheItem object here was passed a parameter. It could be an object that was retrieved from the cache (not a new cacheItem), updating its timestamp here incorrectly "un-expires" the cacheItem. We want to keep the old timestamp if the fetch failed.

If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, remove this line please."_

Reply:
_"The cacheItem object here was passed a parameter. It could be an object that was retrieved from the cache (not a new cacheItem), updating its timestamp here incorrectly "un-expires" the cacheItem. We want to keep the old timestamp if the fetch failed.

If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, remove this line please."_

Reply:
Yeah, I know your optimization for this. That's great!

But currently it's impossible for L2 do the same as L3. Because to judge if the requested locale is supported or not, we need get the supported locale list first like L3, but currently there is no API for getting supported locale list of L2 in Singleton service. So for L2 currently I can't resolve the slowness at first request, only can avoid the slowness in later request by adding an empty CacheItem. 

You're right the cacheItem object maybe retrieved from the cache, but updating its timestamp happens after cacheItem expires, I think it's correct to update the timestamp to un-expires the cacheItem no matter fetch success or failed, then next fetch will happen after cacheItem expired.
But keeping the old timestamp will always make the cacheItem expire and fetch from data source until fetch success which may take up resources.

So I don't agree with you on keeping the old timestamp.



Maybe this can't be done in the next PR/in a short time.

Reply:
Yeah, I know your optimization for this. That's great!

But currently it's impossible for L2 do the same as L3. Because to judge if the requested locale is supported or not, we need get the supported locale list first like L3, but currently there is no API for getting supported locale list of L2 in Singleton service. Maybe this can't be done in the next PR/in a short time. So for L2 currently I can't resolve the slowness at first request, only can avoid the slowness in later request by adding an empty CacheItem.

You're right the cacheItem object maybe retrieved from the cache, but updating its timestamp happens after cacheItem expires, I think it's correct to update the timestamp to un-expires the cacheItem no matter fetch success or failed, then next fetch will happen after cacheItem expired. But keeping the old timestamp will always make the cacheItem expire and fetch from data source until fetch success which may take up resources.

So I don't agree with you on keeping the old timestamp.

Reply:
I now understand why you cannot resolve the slowness on first request. We can think more about it later.

First, your expire logic goes against the agreed upon design. Any design change is outside of the scope of this PR. https://confluence.eng.vmware.com/pages/viewpage.action?pageId=530637182#SingletonClientLibrary-Caching,FallbackMechanism,andOtherFeatures-L10nFetch

Basically, the flaw in your implementation logic is that you would mark a stale cacheItem as not expired, even if you actually weren't able to refresh the data. In other words, the same data that didn't change is actually still expired! Marking it as unexpired is misleading and incorrect. The next request for the same cache item must know that it is actually expired, and therefore try to refresh it.. and yes, it MUST rightfully take up resources to do so. Basically, the client must know whenever it has been using stale data.

Reply:
Hi @huihuiw01 . It looks like you have not filed an issue regarding this yet so I did: https://github.com/vmware/singleton/issues/841

I will now approve this Pr, and you can separately fix the new issue.

Reply:
Thanks for helping log the issue, @jessiejuachon 

But the reason I didn't log the issue is that I don't think my implementation is a issue but a fix solution for below 2 problems exsiting in current design/flow:

Think about such scenario, there are dozens of strings on product's UI page need be translated or formatted, which will call L3 API 'getMessage' or L2 API dozens of times, and at this time Singleton service is down(online mode)

1. When the first fetch failed, the rest call of 'getMessage' will trigger dozens of failed fetch since no cache for failed fetch which may impact performance;

2. When cache expired will trigger dozens of new thread to update cache but since service is down then fetch failed, these failed fetch is resource-intensive. BTW, this problem is tracked by issue #605